Graphics stack and vkmark upgrade

ahsanhussain opened the pull request.

ahsanhussain committed:
meta-flex-staging: Bring in Mesa 24.3.4 and dependencies from BSP layer

Bring in the Mesa upgrade from https://github.com/MentorEmbedded/meta-flex-ref-bsps/pull/274
It has been decided to use this stack across our BSPs to streamline
OSS clearance and support

JIRA-ID: SB-23996

Signed-off-by: Ahsan Hussain <ahsan.hussain@siemens.com>

ahsanhussain committed:
vkmark: Update to v2025.1 for Vulkan headless support

Vulkan support over virtio-gpu is headless at the moment, and for vkmark
benchmark surfaceless Vulkan benchmarking was added in the 2025.1
release. Bring in this change.

JIRA-ID: 24071

Signed-off-by: Ahsan Hussain <ahsan.hussain@siemens.com>

hashraf91 commented on meta-flex-distro/recipes-graphics/vkmark/vkmark/0001-src-meson.build-Prepend-sysroot-for-the-includedir.patch:
```diff
@@ -8,22 +8,23 @@ Need to prepend PKG_CONFIG_SYSROOT_DIR to the dir path for cross compilation
-     output: 'format_map_gen.h',
+ vk_xml = join_paths([
+-    vulkan_dep.get_pkgconfig_variable('prefix'),
++    pkgconf_sysroot, 'usr',
```
better if we use this: `pkgconf_sysroot + vulkan_dep.get_pkgconfig_variable('prefix'),` as it will not override prefix from vulkan_dep and only prepend the pkgconf_sysroot.

ahsanhussain replied:
That's not where the Vulkan registry goes, so it won't resolve to the correct path for the xml

hashraf91 replied:
I have this tested ;). The join_paths functions such that if any of the argument has '/' in front i.e. an absolute path, it discards the previous arguments. If you were joining them with ',' then it was just ignoring the pkgconf_sysroot. It works if you use '+' for concatenation.

ahsanhussain replied:
OK, updated now. Didn't dig into it once the path wasn't resolving :)

hashraf91 commented on meta-flex-staging/recipes-graphics/drm/libdrm_2.4.123.bb:
```diff
@@ -0,0 +1,66 @@
+# This material contains trade secrets or otherwise confidential information owned by Siemens Industry Software Inc.
```
where should the backported recipes be placed? Previously you placed recipes in meta-flex-distro, so I also placed llvm recipe there. And placed only the fix in meta-flex-staging.

ahsanhussain replied:
Ideally in meta-flex-staging, so we can remove from there when we update the Yocto version.

hashraf91 replied:
Okay, you're placing those recipes in meta-flex-staging which are anticipated to be available in future by Yocto update and remaining ones in distro which are just taken from other layers?

ahsanhussain replied:
Kind of. So the recipes anticipated to be updated by the Yocto version update have been going in meta-flex-staging for a while now. That's why we removed a bunch from there when update to scarthgap

ahsanhussain committed:
meta-flex-distro: Add some Vulkan examples

These can be usefull in the future for the BSPs that support the Vulkan
distro feature. These have been ported from kirkstone branch of
upstream meta-vulkan, and updated to remove ties from meta-clang

Ref:

https://github.com/jwinarske/meta-vulkan/blob/kirkstone/recipes-graphics/sascha-samples/sascha-samples_git.bb
https://github.com/jwinarske/meta-vulkan/blob/kirkstone/recipes-graphics/vkcube/vkcube_git.bb

We're not adding the dependency of upstream layer and using recipes from
there because the layer doesn't seem well maintained, and has a
dependency on meta-clang which we don't need

JIRA-ID: SB-24071

Signed-off-by: Ahsan Hussain <ahsan.hussain@siemens.com>

hashraf91 commented on meta-flex-distro/conf/distro/innexis-linux.conf:
```diff
@@ -329,5 +329,11 @@ include conf/distro/include/flex-distro.conf
+# DISTRO preferences
+PREFERRED_VERSION_libdrm = "2.4.123"
+PREFERRED_VERSION_mesa   = "24.3.4"
+PREFERRED_VERSION_wayland-protocols = "1.38"
```
Do we need to force these version? As these will be picked up by default. Their native versions should also be forced to match such as for mesa-native.

ahsanhussain replied:
Unless selected in a high priority layers, these would stay at these versions. If we don't mention here then any recipe from a higher priority layer will override these versions, so it's a good means to have a working configuration in the conf for now. 

As for the native version, do we even care about that? native one doesn't build with any of the new packageconfigs, and also we don't use it so that's why I didn't touch it.

ahsanhussain committed:
innexis-linux: add a small drm_info utility to the RFS

Weston is supported on our BSPs with DRM backend; drm_info is a small
utility to report on the properties of the DRM devices, planes, crtcs
and connectors on the system which is useful to have on the development
image. Add a new recipe for the current latest version of this package

Signed-off-by: Ahsan Hussain <ahsan.hussain@siemens.com>

ahsanhussain commented:
Pushed a commit for drm_info utility.

hashraf91 commented on meta-flex-staging/recipes-graphics/mesa/mesa.inc:
```diff
@@ -0,0 +1,393 @@
+PACKAGECONFIG[v3d] = ""
+PACKAGECONFIG[zink] = ""
+
+GALLIUMDRIVERS = "softpipe,llvmpipe"
```
This line should be split in to following:
```
GALLIUMDRIVERS = "softpipe"
GALLIUMDRIVERS_LLVM:append = ",llvmpipe"
```
because the llvmpipe now gets enabled by gallium-llvm packageconfig.
You can merge the second line with this: https://github.com/MentorEmbedded/meta-sokol-flex/pull/200/files#diff-1bee4e58bd8a7a0a2f2c496a602ca690819bc7435d2d2e43b18883f77e543962R173

ahsanhussain replied:
Keeping the default behavior of this recipe, this should stay here. gallium-llvm packageconfig only used to enable the Vulkan llvmpipe driver (lavapipe) and that's how it is now.

hashraf91 replied:
The behavior has changed. Previously, the recipe was directly depending upon llvm so enabling llvmpipe without any packageconfig was fine but now since the recipe depends on llvm only in case of gallium-llvm packageconfig, so the llvmpipe should only be enabled with that packageconfig.

ahsanhussain replied:
Still depends explicitly on llvm. It was a missing llvm-native which was causing an issue for innexis hybrid BSP build. Thanks for testing this. I've added a commit.

ahsanhussain committed:
mesa: remove llvmpipe from default Gallium drivers

swrast driver was split into softpipe and llvmpipe software rasterizer
drivers, but having llvmpipe requires some extra meson config options;
therefore let's tie this to the gallium-llvm packageconfig and enable
this packageconfig at the distro level

Tested-by: Hasseb Ashraf <haseeb.ashraf@siemens.com>
Signed-off-by: Ahsan Hussain <ahsan.hussain@siemens.com>

ahsanhussain committed:
innexis-linux: Set preferred packageconfig for Vulkan packages

We support Vulkan as a distro feature in headless mode for now.
Configure some of the Vulkan related packages that the user may decide
to build for the same. Some other packages like vkmark/vkcube support
headless system by deafult

JIRA-ID: SB-24075

Signed-off-by: Ahsan Hussain <ahsan.hussain@siemens.com>

ahsanhussain committed:
linux-yocto: fix the license header for the bbappend

This got left out when moving the bbappend from the BSP layer

Signed-off-by: Ahsan Hussain <ahsan.hussain@siemens.com>
